### PR TITLE
Upgrade React

### PR DESCRIPTION
**Pull request by captbaritone**
At the moment I'm aware of one regression. You cannot click the links in the popovers that explain the various plugin installers because they close as soon as you stop hovering over the box.

Review @spicyj

### CODE REVIEW

**Comment by captbaritone** [comment]
I would really love to split this pull request up into smaller pieces, but it looks like each of the various libraries all depend upon having modern versions of each-other. If there is some portion of this can could be extracted into a separate pull request, I would love to be alerted.

File: web/static/js/__tests__/Page-test.jsx [hunk 1]
```diff
@@ -1,17 +1,14 @@
...
 var Page = require('../Page.jsx');
 
 describe('Page', function() {
   it('renders without throwing', function() {
-    var container = document.createElement('div');
-    React.renderComponent(
-      <Page />,
-      container
+    TestUtils.renderIntoDocument(
+      <Page location={{query: ''}} />
```
Comment (captbaritone) [inline web/static/js/__tests__/Page-test.jsx]: Oh yeah, this is an ugly hack. In the real app, `react-router` supplies that property. I should at least add a comment here.

Comment (captbaritone) [inline web/static/js/__tests__/Page-test.jsx]: Updated.

File: conf/webpack.config.js [hunk 2]
```diff
@@ -31,7 +31,13 @@ module.exports = {
   devtool: 'source-map',
   module: {
     loaders: [
-      { test: /\.jsx$/, loader: "jsx-loader?harmony" }
+      {
+        test: /\.jsx$/,
+        loader: "babel",
+        query: {
+          presets: ['react']
```
Comment (spicyj) [inline conf/webpack.config.js]: babel-loader should pick up .babelrc automatically so hopefully you don't need this.

File: web/static/js/Plugin.jsx [hunk 3]
```diff
@@ -23,22 +22,22 @@ var Plugin = React.createClass({
...
-    $(this.getDOMNode()).find('[title]').tooltip({
+    $(findDOMNode(this)).find('[title]').tooltip({
       animation: false,
       container: 'body'
     });
   },
 
   goToDetailsPage: function() {
-    transitionTo("plugin", {slug: this.props.plugin.slug});
+    browserHistory.push("plugin/" + this.props.plugin.slug);
```
Comment (spicyj) [inline web/static/js/Plugin.jsx]: Woah, you need to construct URLs manually now?

File: web/static/js/__tests__/Page-test.jsx [hunk 4]
```diff
@@ -1,17 +1,15 @@
...
 var Page = require('../Page.jsx');
 
 describe('Page', function() {
   it('renders without throwing', function() {
-    var container = document.createElement('div');
-    React.renderComponent(
-      <Page />,
-      container
+    TestUtils.renderIntoDocument(
+      // Normallsyg react-routes adds a location. For tests we do it ourselves.
```
Comment (spicyj) [inline web/static/js/__tests__/Page-test.jsx]: "Normallsyg"

File: web/static/js/app.jsx [hunk 5]
```diff
@@ -885,18 +884,22 @@ var PluginListPage = React.createClass({
...
-var routes = <Route handler={Page} location="history">
-  <Route name="plugin-list" path="/" handler={PluginListPage} />
-  <Route name="plugin" path="plugin/:slug" handler={PluginPage} />
-  <Route name="submit" handler={SubmitPage} />
-  <Route name="thanks-for-submitting" handler={ThanksForSubmittingPage} />
-  <Route name="about" handler={AboutPage} />
-  <Route path="*" handler={NotFound} />
-</Route>;
-React.renderComponent(routes, document.body);
+render((
```
Comment (spicyj) [inline web/static/js/app.jsx]: (Inner parens here are unnecessary.)

**Comment by spicyj** [comment]
Neat! Thanks.

File: web/static/js/Plugin.jsx (hunk 3 above)
Comment (captbaritone) [inline web/static/js/Plugin.jsx]: It is entirely possible that I'm missing something. From what I could find, this is expected to either be a string (path), or a [`LocationDescription`](https://github.com/mjackson/history/blob/master/docs/Glossary.md#locationdescriptor), and I didn't see any way to declare route wildcards in a LocationDescription.

**Pull request merged**